caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -309,71 +309,111 @@ public boolean isFile( ConcreteResource resource )
     public InputStream openInputStream( final ConcreteResource resource )
             throws IOException
     {
-        //use "this" as lock is heavy, should think about use the transfer for the resource as the lock for each thread
-        synchronized ( getTransfer( resource ) )
-        {
-            boolean localExisted = plCacheProvider.exists( resource );
-            if ( !localExisted )
+        final String pathKey = getKeyForResource( resource );
+
+        // This lock is used to control the the local resource can be opened successfully finally when local resource missing
+        // but NFS not, which means will do a NFS->local copy.
+        final Object copyLock = new Object();
+        // A flag to mark if the local resource can be open now or need to wait for the copy thread completes its work
+        final AtomicBoolean canStreamOpen = new AtomicBoolean( false );
+        // This copy task is responsible for the NFS->local copy, and will be run in another thread,
+        // which can use PartyLine concurrent read/write function on the local cache to boost
+        // the i/o operation
+        final Runnable copyNFSTask = new Runnable()
+        {
+            @Override
+            public void run()
             {
-                final String pathKey = getKeyForResource( resource );
-                // will run the NFS->local copy in another thread, which can use PartyLine concurrent read/write function
-                // on the local cache to boost the i/o operation
-                executor.execute( new Runnable()
+                InputStream nfsIn = null;
+                OutputStream localOut = null;
+                synchronized ( copyLock )
```

Comment:
I'm not sure if it matters, but I think you can wait to synchronize on the copyLock until you're ready to notify it. The other side, wait() will also need that synchronized block, but I'm pretty sure this whole section doesn't need to be synchronized that way (since the other side, of opening the input stream for the user, will wait until copyLock is notified).

This does bring up an interesting point, though. If another thread starts the openInputStream() process right after this one starts, but before the thread starts, it may try to spawn a second writer thread. I'm not sure whether this will result in an error, or corrupted data, or just one of the calls waiting longer for the first copy/read to complete.

I haven't had enough time to read up on ReentrantLock yet, but I'm wondering if maintaining a Map<String, ReentrantLock> field at the instance level might help here. We could use that in place of copyLock, maybe? I'm not 100% sure that makes sense. But it does feel like we're setting up a race condition here. I guess maybe a test would be a reasonable place to start pursuing that suspicion... :-\